scilpy/tractograms/tractogram_operations.py

```diff
@@ -843,3 +854,246 @@ def split_sft_randomly_per_cluster(orig_sft, chunk_sizes, seed, thresholds):
+            progress_bar.update(len(chunk))
+
+    return diff_data, acc_data
+
+
+def tractogram_pairwise_comparison(sft_one, sft_two, mask, nbr_cpu=1,
+                                   skip_streamlines_distance=True):
+    """
+    Compute the difference between two sets of streamlines for each voxel in
+    the mask. This function uses multiprocessing to compute the difference
+    between two sets of streamlines for each voxel.
+
+    Parameters
+    ----------
+    sft_one: StatefulTractogram
+        First tractogram to compare.
+    sft_two: StatefulTractogram
+        Second tractogram to compare.
+    mask: np.ndarray
+        Mask of the data to compare (optional).
+    nbr_cpu: int
+        Number of CPU to use (default: 1).
+    skip_streamlines_distance: bool
+        If true, skip the computation of the distance between streamlines.
+        (default: True)
+
+    Returns
+    -------
+    List of np.ndarray
+        acc_norm: Angular correlation coefficient.
+        corr_norm: Correlation coefficient of density maps.
+        diff_norm: Voxelwise distance between sets of streamlines.
+        heatmap: Merged heatmap of the three metrics using harmonic mean.
+    """
+    global sft_1, sft_2
+    sft_1, sft_2 = sft_one, sft_two
+
+    sft_1.to_vox()
+    sft_2.to_vox()
+    sft_1.streamlines._data = sft_1.streamlines._data.astype(np.float16)
+    sft_2.streamlines._data = sft_2.streamlines._data.astype(np.float16)
+    dimensions = tuple(sft_1.dimensions)
+
+    global matched_points_1, matched_points_2
+    matched_points_1 = generate_matched_points(sft_1)
+    matched_points_2 = generate_matched_points(sft_2)
+
+    logging.info('Computing KDTree...')
+    global tree_1, tree_2
+    tree_1 = cKDTree(sft_1.streamlines._data)
+    tree_2 = cKDTree(sft_2.streamlines._data)
+
+    # Limits computation to mask AND streamlines (using density)
+    if not mask:
+        mask = np.ones(dimensions)
+
+    logging.info('Computing density maps...')
+    density_1 = compute_tract_counts_map(sft_1.streamlines,
+                                         dimensions).astype(float)
+    density_2 = compute_tract_counts_map(sft_2.streamlines,
+                                         dimensions).astype(float)
+    mask = density_1 * density_2 * mask
+    mask[mask > 0] = 1
+
+    logging.info('Computing correlation map...')
+    corr_data = correlation([density_1, density_2], None) * mask
+
+    logging.info('Computing TODI #1...')
+    global sh_data_1, sh_data_2
+    sft_1.to_corner()
+    todi_obj = TrackOrientationDensityImaging(dimensions, 'repulsion724')
+    todi_obj.compute_todi(deepcopy(sft_1.streamlines), length_weights=True)
+    todi_obj.mask_todi(mask)
+    sh_data_1 = todi_obj.get_sh('descoteaux07', 8)
+    sh_data_1 = todi_obj.reshape_to_3d(sh_data_1)
+    sft_1.to_center()
+
+    logging.info('Computing TODI #2...')
+    sft_2.to_corner()
+    todi_obj = TrackOrientationDensityImaging(dimensions, 'repulsion724')
```

Comment:
Eventually, we should have a method 'compute_todi' that would also be used in scil_tractogram_compute_todi. If you don't do it here, let's add a todo in our google doc or on github.